codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java

```diff
@@ -187,16 +187,22 @@ public void operationComplete(ChannelFuture future) throws Exception {
                 }
                 ChannelFuture future = frameWriter.writeHeaders(ctx, streamId, headers, streamDependency,
                                                                 weight, exclusive, padding, endOfStream, promise);
-                // Synchronously set the headersSent flag to ensure that we do not subsequently write
-                // other headers containing pseudo-header fields.
-                stream.headersSent();
+                // Writing headers may fail during the encode state if they violate HPACK limits.
+                Throwable failureCause = future.cause();
```

Comment:
Sounds a bit like we depend on an implementation detail here. I guess for now it's ok but definitely something we should revisit later on.